hip-584 evm module (#584)
hashgraph/hedera-improvement-proposal: HIP/hip-584.md

In order to be possible to both execute and simulate transactions, run by the EVM used inside `hedera-node`,
we need to extract a module encapsulating all of the EVM logic.

The resulting library would allow us to fulfil the following use cases:

- Execute transactions in a regular flow with committing changes to state
- Execute transaction simulations that perform read-only operations
- Execute transaction simulations that perform speculative writes without committing changes to state
- Execute transaction simulations that perform speculative writes using historical data for transaction replay, 
without committing changes to state

## User stories

- As a developer, I want to have a library with common Hedera smart contract EVM logic that I can reuse in different projects
- As a user, I want to perform `eth_call` calls from the Ethereum JSON-RPC standard
- As a user, I want to perform `eth_gasEstimate` calls from the Ethereum JSON-RPC standard

## Specification

### EVM Library Definition

We want the EVM library to be used by multiple clients like the `hedera-services` and `hedera-mirror-node` repos
and to be able to be extended for a future use from additional clients.

> To facilitate differences, we need to define interfaces, which would be inside the library 

Reviewer: Should we not be listing the suggested major interfaces and their responsibilities here?

Reviewer 2: We may define the basic interface for interacting with the library. It will be inside `hedera-evm-api` module and `mirror-node` would use it to spin up a simulation of a transaction. The output would be `TransactionProcessingResult`. `hedera-evm-api` on its own side would consume and use the `hedera-evm` implementation module to perform the transaction. The dependency chain would look like this:
`besu` -> `hedera-evm` -> `hedera-evm-api `-> `hedera-node`/`hedera-mirror-web3`

Going further into more details about the other interfaces, IMO, might introduce more specific and unnecessary information for the context of a HIP.

Reviewer: Maybe this can be in a design doc committed into the services repo inside the `hedera-evm` repo

and each client of the library (in this case `hedera-node` and `mirror-node`) would provide its own implementation.

### Mirror Node

Information will be made available via new REST API

#### Read-only queries and gas estimate REST API

> A new  `/api/v1/contracts/call` REST API will be added to execute read-only queries and gas estimate.

Reviewer: Note that this will be a `POST` since all current APIs are `GET`

}
```

## Backwards Compatibility

Since the library would duplicate some code inside `hedera-node`, there would be a need to extract
and remove this code from `hedera-node` and start using `hedera-evm` as a dependency inside `hedera-node`.
In this way `hedera-node` and `mirror-node` would reuse the same code base, and we would avoid code duplications.

`Mirror Nodes` that will support the new REST APIs should have enabled importing of `CONTRACT_BYTECODE` and `CONTRACT_STATE_CHANGE` 
sidecar types. Otherwise, executing the endpoints would result in missing runtime bytecode for execution
or using stale contract storage data.

## Security Implications

There would be some security implementations for `Mirror Nodes`. Since the invocation of `eth_call` and `eth_gasEstimate` RPC calls 
would be free of charge and the `Mirror Nodes` don’t have throttle mechanism, some attack vectors would be possible. 
The Nodes could be overloaded with huge amount of calls or calls invoking smart contract methods with huge gas usage, 
both of which might take more system usage and slow down the network for other users. A rate limit mechanism will be implemented,

> so that the load put on Mirror Nodes to be balanced.

Reviewer 2: Mirror Nodes -> `Mirror Nodes`

## How to Teach This

> Respective documentation will be added.

Reviewer: Can be more specific with
- Service "EVM Modularization" Design Doc
- "Mirror EVM Archive Node" Design Doc
- Open API doc on Mirror Node
- docs.hedera.com sections for `/api/v1/contracts/call`

Reviewer 2: I believe we can link these corresponding documentations at a later point, so that they would more up-to-date with the latest implementation details for the evm modules.

---
hip: 584
title: EVM Module
author:  Ivan Kavaldzhiev <ivan.kavaldzhiev@limechain.tech>, Mustafa Uzun <mustafa.uzun@limechain.tech>
type: Standards Track
category: Service
needs-council-approval: Yes
status: Review
created: 2022-09-27
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/586
updated: 2022-10-04
---

## Abstract

> Encapsulating Hedera smart contract EVM specific logic in a library.
> It would expose interfaces that will allow different implementations.
> Mirror node would build ethereum-related REST API on top of the library. 
> Mirror nodes could use the library to simulate a transaction and the output would be used to 
> populate the return value for the REST API that is invoked. 
> Consensus nodes would also use the library and keep their existing logic by providing specific interface implementations.

Reviewer 3: ```suggestion
This HIP describes the interfaces and modules that will allow the EVM implementation to be separated into a Java library. This will enable a number of future features such as different EVM implementations, EVM execution APIs on the mirror node, and transaction simulation and verification off of the consensus nodes.  The consensus nodes would continue to use this library.
```

Reviewer 2: @shemnon, @Nana-EC  I think that saying `This HIP describes the interfaces and modules` would imply further and more detailed definitions about the implementation details for the EVM module. I think we agreed on being more abstract in terms of explanations in this HIP and if needed, link a more detailed design doc with interface definitions and examples at a later point.

## Motivation

> Extracting common Hedera smart contract EVM logic in a single place

Reviewer 3: The motivation section should be prose not a bullet list.

```suggestion
Currently the EVM code is deeply embedded in consensus node logic.  However ongoing user experience work needs EVM execution to enable features such as cost-free execution of read-only smart contract queries, gas estimation, and transient simulation of read-write operations.

In order to maximize compatibility and reduce defects it would be best to run all these activities from the same codebase.  In order to do so a library needs to be extracted from the consensus logic and the consensus nodes and external programs such as the mirror node.
```

Enabling free read-only queries.

Enabling free gas estimate for a given transaction.

Enabling free transaction simulations.

## Rationale

In order to be possible to both execute and simulate transactions, run by the EVM used inside `hedera-node`,
we need to extract a module encapsulating all of the EVM logic.

The resulting library would allow us to fulfil the following use cases:

- Execute transactions in a regular flow with committing changes to state
- Execute transaction in a regular flow or simulations that perform read-only operations
- Execute transaction simulations that perform speculative writes without committing changes to state
- Execute transaction simulations that perform speculative writes using historical data for transaction replay, 
without committing changes to state

## User stories

- As a developer, I want to have a library with common Hedera smart contract EVM logic that I can reuse in different projects
- As a user, I want to perform `eth_call` calls from the Ethereum JSON-RPC standard
- As a user, I want to perform `eth_gasEstimate` calls from the Ethereum JSON-RPC standard

> 

Reviewer 3: I think we need a story for the debug use case, worded so it's just the enabling logic and not the full debug support

```suggestion
- As a user, I want a library that can be used to re-run historical EVM transactions accurately so that I can inspect and debug the execution of the bytecode logic.

```
